src/core/WakaTimeCore.ts

```diff
@@ -38,13 +48,211 @@ class WakaTimeCore {
+        .trim()
+        .replace(/(\/|@@)$/, '')
+        .replace(/^(?:https?:\/\/)?/i, '');
+      if (cleanLine === '') continue;
+
+      const projectIndicatorIndex = cleanLine.lastIndexOf(projectIndicatorCharacters);
+      const projectIndicatorExists = projectIndicatorIndex > -1;
+      let projectName = null;
+      let urlFromLine = cleanLine;
+      if (projectIndicatorExists) {
+        const start = projectIndicatorIndex + projectIndicatorCharacters.length;
+        projectName = cleanLine.substring(start);
+        urlFromLine = cleanLine
+          .replace(cleanLine.substring(projectIndicatorIndex), '')
+          .replace(/\/$/, '');
+      }
+      const schemaHttpExists = url.match(/^http:\/\//i);
+      const schemaHttpsExists = url.match(/^https:\/\//i);
+      let schema = '';
+      if (schemaHttpExists) {
+        schema = 'http://';
+      }
+      if (schemaHttpsExists) {
+        schema = 'https://';
+      }
+      const cleanUrl = url
+        .trim()
+        .replace(/(\/|@@)$/, '')
+        .replace(/^(?:https?:\/\/)?/i, '');
+      const startsWithUrl = cleanUrl.toLowerCase().includes(urlFromLine.toLowerCase());
+      if (startsWithUrl) {
+        return {
+          project: projectName,
+          url: schema + urlFromLine,
+        };
+      }
+    }
+
+    return {
+      project: null,
+      url: null,
+    };
+  }
+
+  /**
+   * Given the heartbeat and logging type it creates a payload and
+   * sends an ajax post request to the API.
+   *
+   * @param heartbeat
+   * @param debug
+   */
+  async sendHeartbeat(heartbeat: SendHeartbeat, debug: boolean, apiKey: string): Promise<void> {
+    let payload;
+
+    const loggingType = await this.getLoggingType();
+    // Get only the domain from the entity.
+    // And send that in heartbeat
+    if (loggingType == 'domain') {
+      heartbeat.url = getDomainFromUrl(heartbeat.url);
+      payload = this.preparePayload(heartbeat, 'domain', debug);
+      await this.sendPostRequestToApi(payload, apiKey);
+    }
+    // Send entity in heartbeat
+    else if (loggingType == 'url') {
+      payload = this.preparePayload(heartbeat, 'url', debug);
+      await this.sendPostRequestToApi(payload, apiKey);
+    }
+  }
+
+  /**
+   * Returns a promise with logging type variable.
+   *
+   * @returns {*}
+   * @private
+   */
+  async getLoggingType(): Promise<string> {
+    const items = await browser.storage.sync.get({
+      loggingType: config.loggingType,
+    });
+
+    return items.loggingType;
+  }
+
+  /**
+   * Creates payload for the heartbeat and returns it as JSON.
+   *
+   * @param heartbeat
+   * @param type
+   * @param debug
+   * @returns {*}
+   * @private
+   */
+  preparePayload(heartbeat: SendHeartbeat, type: string, debug = false): Record<string, unknown> {
+    return {
+      entity: heartbeat.url,
+      is_debugging: debug,
+      plugin: 'browser-wakatime/' + config.version,
+      project: heartbeat.project ?? '<<LAST_PROJECT>>',
```

Comment:
Should not send `<<LAST_PROJECT>>` for empty ones? Instead an empty string would be sent?

Reply:
Everytime i send post request with `<<LAST_PROJECT>>` it does not get logged in the dashboard, we could skip this field if project is not present since it is optional `"project": <string: project name (optional)>,`

Reply:
I beleieve skipping it makes more sense to me and we've been doing it in other IDEs.